contracts/index_executor/src/task.rs

```diff
@@ -257,6 +321,90 @@ impl Task {
         Ok(())
     }
 
+    pub fn apply_recipient(&mut self, context: &Context) -> Result<(), &'static str> {
+        let step_count = self.steps.len();
+        for (index, step) in self.steps.iter_mut().enumerate() {
+            let step_source_chain = &step.source_chain(context).ok_or("MissingChain")?;
+            let step_dest_chain = &step.dest_chain(context).ok_or("MissingChain")?;
+            let worker_account_info = context.get_account(self.worker).ok_or("WorkerNotFound")?;
+
+            // For sure last step we should put real recipient, or else the recipient could be either
+            // worker account or handler account
+            step.recipient = if index == step_count - 1 {
+                Some(self.recipient.clone())
+            } else {
+                // If bridge to a EVM chain, asset should be send to Handler account to execute the reset of calls
+                if step.is_bridge_step() {
+                    if step_dest_chain.is_evm_chain() {
+                        Some(step_dest_chain.handler_contract.clone())
+                    } else {
+                        Some(worker_account_info.account32.to_vec())
+                    }
+                } else {
+                    // For non-bridge operatoions, because we don't batch call in Sub chains, so recipient should
+                    // be worker account on source chain, or should be Handler address on source chain
+                    if step_source_chain.is_sub_chain() {
+                        let worker_account = match step_source_chain.chain_type {
+                            ChainType::Evm => worker_account_info.account20.to_vec(),
+                            ChainType::Sub => worker_account_info.account32.to_vec(),
+                        };
+                        Some(worker_account)
+                    } else {
+                        Some(step_source_chain.handler_contract.clone())
+                    }
+                }
+            };
+        }
+        Ok(())
+    }
+
+    pub fn merge_step(&mut self, context: &Context) -> Result<(), &'static str> {
```

Comment:
Haven't fully review it. Would be nice if you can add some comments to describe how it works.

Reply:
TBH, I don't see there are a lot of improvements except less if .. else .. patterns. We are moving merge to SDK BTW